skfp/fingerprints/lingo.py

```diff
@@ -0,0 +1,183 @@
+        J. Chem. Inf. Model. 2005, 45, 2, 386-393
+        <https://pubs.acs.org/doi/abs/10.1021/ci0496797>`_
+
+    Examples
+    --------
+    >>> from skfp.fingerprints import LingoFingerprint
+    >>> smiles = ["[C-]#N", "CC(=O)NC1=CC=C(C=C1)O"]
+    >>> fp = LingoFingerprint()
+    >>> fp
+    LingoFingerprint()
+    >>> fp.transform(smiles)
+    array([[0, 0, 0, ..., 0, 0, 0],
+       [0, 1, 0, ..., 0, 0, 0]], dtype=uint8)
+    """
+
+    _parameter_constraints: dict = {
+        **FingerprintTransformer._parameter_constraints,
+        "fp_size": [Interval(Integral, 1, None, closed="left")],
+        "substring_length": [Interval(Integral, 1, None, closed="left")],
+        "count": ["boolean"],
+    }
+
+    def __init__(
+        self,
+        fp_size: int = 1024,
+        substring_length: int = 4,
+        count: bool = False,
+        sparse: bool = False,
+        n_jobs: Optional[int] = None,
+        batch_size: Optional[int] = None,
+        verbose: int = 0,
+    ):
+        super().__init__(
+            n_features_out=fp_size,
+            sparse=sparse,
+            n_jobs=n_jobs,
+            batch_size=batch_size,
+            verbose=verbose,
+        )
+        self.fp_size = fp_size
+        self.substring_length = substring_length
+        self.count = count
+
+    def transform(
+        self, X: Sequence[Union[str, Mol]], copy: bool = False
+    ) -> Union[np.ndarray, csr_array]:
+        """
+        Compute Lingo fingerprints.
+
+        Parameters
+        ----------
+        X : {sequence, array-like} of shape (n_samples,)
+            Sequence containing SMILES strings or RDKit Mol objects.
+
+        copy : bool, default=False
+            Whether to copy X or modify it in place.
+
+        Returns
+        -------
+        X: {ndarray, sparse matrix} of shape (n_samples, self.fp_size)
+            2D array containing Lingo fingerprints.
+        """
+        return super().transform(X, copy)
+
+    def smiles_to_dicts(
+        self, X: Sequence[Union[str, Mol]]
+    ) -> list[defaultdict[str, int]]:
+        """
+        Convert SMILES strings to dictionaries of substring counts.
+
+        Parameters
+        ----------
+        X: {sequence, array-like} of shape (n_samples,)
+            Sequence containing SMILES strings or RDKit Mol objects.
+
+        Returns
+        -------
+        result: list[defaultdict[str, int]]
+            List of dictionaries containing substring counts.
+        """
+        X = ensure_smiles(X)
+        X = [re.sub(r"[123456789]", "0", x) for x in X]
+        X = [re.sub(r"Cl", "L", x) for x in X]
+        X = [re.sub(r"Br", "R", x) for x in X]
+
+        result = []
+        for smiles in X:
+            result_dict: defaultdict[str, int] = defaultdict(int)
+            for i in range(len(smiles) - self.substring_length + 1):
+                result_dict[smiles[i : i + self.substring_length]] += 1
+            result.append(result_dict)
```

Comment:
Consider an alternative solution:
```
substrings = [smiles[i : i + self.substring_length] for i in range(len(smiles) - self.substring_length + 1)]
substring_counts = dict(Counter(substrings))
```
Which one to use should be based on speed

Reply:
I have done performance tests on both solutions. I do not see a significant gain (or any difference) from changing solutions. Both solutions transform hiv dataset locally in ~3.7-3.9s.

Reply:
I would leave the original solution. I think it's easier to read